// n8n-nodes-base/dist/credentials/CarbonBlackApi.credentials.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.CarbonBlackApi = void 0;
class CarbonBlackApi {
    constructor() {
        this.name = 'carbonBlackApi';
        this.displayName = 'Carbon Black API';
        this.icon = 'file:icons/vmware.svg';
        this.documentationUrl = 'carbonblack';
        this.httpRequestNode = {
            name: 'Carbon Black',
            docsUrl: 'https://developer.carbonblack.com/reference',
            apiBaseUrl: '',
        };
        this.properties = [
            {
                displayName: 'URL',
                name: 'apiUrl',
                type: 'string',
                placeholder: 'https://defense.conferdeploy.net/',
                default: '',
            },
            {
                displayName: 'Access Token',
                name: 'accessToken',
                type: 'string',
                typeOptions: { password: true },
                required: true,
                default: '',
            },
        ];
        this.authenticate = {
            type: 'generic',
            properties: {
                headers: {
                    'X-Auth-Token': '={{$credentials.accessToken}}',
                },
            },
        };
    }
}
exports.CarbonBlackApi = CarbonBlackApi;
