// @n8n/n8n-nodes-langchain/dist/nodes/Guardrails/Guardrails.node.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var Guardrails_node_exports = {};
__export(Guardrails_node_exports, {
  Guardrails: () => Guardrails
});
module.exports = __toCommonJS(Guardrails_node_exports);
var import_n8n_workflow = require("n8n-workflow");
var import_GuardrailsV1 = require("./v1/GuardrailsV1.node");
var import_GuardrailsV2 = require("./v2/GuardrailsV2.node");
class Guardrails extends import_n8n_workflow.VersionedNodeType {
  constructor() {
    const baseDescription = {
      displayName: "Guardrails",
      name: "guardrails",
      icon: "file:guardrails.svg",
      group: ["transform"],
      defaultVersion: 2,
      description: "Safeguard AI models from malicious input or prevent them from generating undesirable responses",
      codex: {
        alias: ["LangChain", "Guardrails", "PII", "Secret", "Injection", "Sanitize"],
        categories: ["AI"],
        subcategories: {
          AI: ["Agents", "Miscellaneous", "Root Nodes"]
        },
        resources: {
          primaryDocumentation: [
            {
              url: "https://docs.n8n.io/integrations/builtin/core-nodes/n8n-nodes-langchain.guardrails/"
            }
          ]
        }
      }
    };
    const nodeVersions = {
      1: new import_GuardrailsV1.GuardrailsV1(baseDescription),
      2: new import_GuardrailsV2.GuardrailsV2(baseDescription)
    };
    super(nodeVersions, baseDescription);
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  Guardrails
});
