// @n8n/n8n-nodes-langchain/dist/utils/agent-execution/index.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var agent_execution_exports = {};
__export(agent_execution_exports, {
  buildSteps: () => import_buildSteps.buildSteps,
  buildToolContext: () => import_memoryManagement.buildToolContext,
  createEngineRequests: () => import_createEngineRequests.createEngineRequests,
  loadMemory: () => import_memoryManagement.loadMemory,
  processEventStream: () => import_processEventStream.processEventStream,
  saveToMemory: () => import_memoryManagement.saveToMemory
});
module.exports = __toCommonJS(agent_execution_exports);
var import_createEngineRequests = require("./createEngineRequests");
var import_buildSteps = require("./buildSteps");
var import_processEventStream = require("./processEventStream");
var import_memoryManagement = require("./memoryManagement");
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  buildSteps,
  buildToolContext,
  createEngineRequests,
  loadMemory,
  processEventStream,
  saveToMemory
});
