// @n8n/n8n-nodes-langchain/dist/utils/agent-execution/memoryManagement.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var memoryManagement_exports = {};
__export(memoryManagement_exports, {
  buildToolContext: () => buildToolContext,
  loadMemory: () => loadMemory,
  saveToMemory: () => saveToMemory
});
module.exports = __toCommonJS(memoryManagement_exports);
var import_messages = require("@langchain/core/messages");
function buildToolContext(steps) {
  return steps.map(
    (step) => `Tool: ${step.action.tool}, Input: ${JSON.stringify(step.action.toolInput)}, Result: ${step.observation}`
  ).join("; ");
}
async function loadMemory(memory, model, maxTokens) {
  if (!memory) {
    return void 0;
  }
  const memoryVariables = await memory.loadMemoryVariables({});
  let chatHistory = memoryVariables["chat_history"] || [];
  if (maxTokens && model) {
    chatHistory = await (0, import_messages.trimMessages)(chatHistory, {
      strategy: "last",
      maxTokens,
      tokenCounter: model,
      includeSystem: true,
      startOn: "human",
      allowPartial: true
    });
  }
  return chatHistory;
}
async function saveToMemory(input, output, memory, steps) {
  if (!output || !memory) {
    return;
  }
  let fullOutput = output;
  if (steps && steps.length > 0) {
    const toolContext = buildToolContext(steps);
    fullOutput = `[Used tools: ${toolContext}] ${fullOutput}`;
  }
  await memory.saveContext({ input }, { output: fullOutput });
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  buildToolContext,
  loadMemory,
  saveToMemory
});
