// @n8n/n8n-nodes-langchain/dist/credentials/AzureAiSearchApi.credentials.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var AzureAiSearchApi_credentials_exports = {};
__export(AzureAiSearchApi_credentials_exports, {
  AzureAiSearchApi: () => AzureAiSearchApi
});
module.exports = __toCommonJS(AzureAiSearchApi_credentials_exports);
class AzureAiSearchApi {
  constructor() {
    this.name = "azureAiSearchApi";
    this.displayName = "Azure AI Search API";
    this.documentationUrl = "azureaisearch";
    this.properties = [
      {
        displayName: "Search Endpoint",
        name: "endpoint",
        type: "string",
        required: true,
        default: "",
        placeholder: "https://your-search-service.search.windows.net"
      },
      {
        displayName: "API Key",
        name: "apiKey",
        type: "string",
        typeOptions: { password: true },
        required: true,
        default: ""
      }
    ];
    this.authenticate = async (credentials, requestOptions) => {
      return {
        ...requestOptions,
        headers: {
          ...requestOptions.headers,
          "api-key": credentials.apiKey
        }
      };
    };
    this.test = {
      request: {
        baseURL: "={{$credentials.endpoint}}/indexes",
        url: "",
        method: "GET",
        headers: {
          accept: "application/json"
        },
        qs: {
          "api-version": "2024-07-01"
        }
      }
    };
  }
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  AzureAiSearchApi
});
