// @n8n/n8n-nodes-langchain/dist/nodes/mcp/McpClient/listSearch.js
"use strict";
var __defProp = Object.defineProperty;
var __getOwnPropDesc = Object.getOwnPropertyDescriptor;
var __getOwnPropNames = Object.getOwnPropertyNames;
var __hasOwnProp = Object.prototype.hasOwnProperty;
var __export = (target, all) => {
  for (var name in all)
    __defProp(target, name, { get: all[name], enumerable: true });
};
var __copyProps = (to, from, except, desc) => {
  if (from && typeof from === "object" || typeof from === "function") {
    for (let key of __getOwnPropNames(from))
      if (!__hasOwnProp.call(to, key) && key !== except)
        __defProp(to, key, { get: () => from[key], enumerable: !(desc = __getOwnPropDesc(from, key)) || desc.enumerable });
  }
  return to;
};
var __toCommonJS = (mod) => __copyProps(__defProp({}, "__esModule", { value: true }), mod);
var listSearch_exports = {};
__export(listSearch_exports, {
  getTools: () => getTools
});
module.exports = __toCommonJS(listSearch_exports);
var import_utils = require("../shared/utils");
async function getTools(filter, paginationToken) {
  const authentication = this.getNodeParameter("authentication");
  const serverTransport = this.getNodeParameter("serverTransport");
  const endpointUrl = this.getNodeParameter("endpointUrl");
  const node = this.getNode();
  const { headers } = await (0, import_utils.getAuthHeaders)(this, authentication);
  const client = await (0, import_utils.connectMcpClient)({
    serverTransport,
    endpointUrl,
    headers,
    name: node.type,
    version: node.typeVersion,
    onUnauthorized: async (headers2) => await (0, import_utils.tryRefreshOAuth2Token)(this, authentication, headers2)
  });
  if (!client.ok) {
    throw (0, import_utils.mapToNodeOperationError)(node, client.error);
  }
  const result = await client.result.listTools({ cursor: paginationToken });
  const tools = filter ? result.tools.filter((tool) => tool.name.toLowerCase().includes(filter.toLowerCase())) : result.tools;
  return {
    results: tools.map((tool) => ({
      name: tool.name,
      value: tool.name,
      description: tool.description,
      inputSchema: tool.inputSchema
    })),
    paginationToken: result.nextCursor
  };
}
// Annotate the CommonJS export names for ESM import in node:
0 && (module.exports = {
  getTools
});
