// n8n/dist/commands/mfa/disable.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.DisableMFACommand = void 0;
const command_1 = require("@oclif/command");
const BaseCommand_1 = require("../BaseCommand");
const typedi_1 = __importDefault(require("typedi"));
const user_repository_1 = require("../../databases/repositories/user.repository");
class DisableMFACommand extends BaseCommand_1.BaseCommand {
    async init() {
        await super.init();
    }
    async run() {
        const { flags } = this.parse(DisableMFACommand);
        if (!flags.email) {
            this.logger.info('An email with --email must be provided');
            return;
        }
        const updateOperationResult = await typedi_1.default.get(user_repository_1.UserRepository).update({ email: flags.email }, { mfaSecret: null, mfaRecoveryCodes: [], mfaEnabled: false });
        if (!updateOperationResult.affected) {
            this.reportUserDoesNotExistError(flags.email);
            return;
        }
        this.reportSuccess(flags.email);
    }
    async catch(error) {
        this.logger.error('An error occurred while disabling MFA in account');
        this.logger.error(error.message);
    }
    reportSuccess(email) {
        this.logger.info(`Successfully disabled MFA for user with email: ${email}`);
    }
    reportUserDoesNotExistError(email) {
        this.logger.info(`User with email: ${email} does not exist`);
    }
}
exports.DisableMFACommand = DisableMFACommand;
DisableMFACommand.description = 'Disable MFA authentication for a user';
DisableMFACommand.examples = ['$ n8n mfa:disable --email=johndoe@example.com'];
DisableMFACommand.flags = {
    help: command_1.flags.help({ char: 'h' }),
    email: command_1.flags.string({
        description: 'The email of the user to disable the MFA authentication',
    }),
};
